dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java

```diff
@@ -321,6 +354,29 @@ public String toString() {
         }
     }
 
+    /**
+     * Checks if there is remaining data in the stream, note that this is a
+     * destructive operation which should only occur in order to close the stream.
+     */
+    private static boolean hasSubstantialRemainingData(CloseableHttpResponse response) {
+        try {
+            HttpEntity entity = response.getEntity();
+            if (entity == null || !entity.isStreaming()) {
+                return false;
+            }
+            InputStream stream = entity.getContent();
+            // Fast check: The stream has been fully exhausted in the expected case,
+            // no need to create buffers for drainage unless we know there's data to drain.
+            if (stream.read() == -1) {
+                return false;
+            }
+            return REMAINING_CONTENT_CONNECTION_DISCARD_THRESHOLD
+                    == ByteStreams.exhaust(ByteStreams.limit(stream, REMAINING_CONTENT_CONNECTION_DISCARD_THRESHOLD));
```

Comment:
So this naturally allocates a buffer just for this check: is there a way we can do this without allocating? Also we allocate 2 implementations of an InputStream

Reply:
That's correct, however in the expected case the `stream.read() == -1` check on line 370 avoids this path entirely because the stream has been fully exhausted. For what it's worth, the call to `close()` in both `ChunkedInputStream` and `ContentLengthInputStream` (from hc) allocate new buffers in order to drain remaining contents, so this shouldn't be dramatically worse.